Multiple AI fixes

13spacemen opened the pull request:
Fixes https://github.com/tgstation/tgstation/issues/68457
Fixes https://github.com/tgstation/tgstation/issues/68468

Also fixes an oversight in that you can't (un)anchor AIs with a wrench, you can on Baycode and I assumed it was like that on TG already

Also minor code edit, added an arg to skip MMI creation on AI mob deconstruction since the MMI created gets overwritten anyway

:cl:
fix: Carding an AI or deconstructing it will not ping ghosts about its positronic brain
fix: Ejecting a brain from an MMI will correctly transfer the guy in the MMI to the ejected brain
fix: You can (un)anchor AIs with a wrench
/:cl:

13spacemen committed:
AI (de)construction posibrains will not ping ghosts

Mooshimi commented:
what does this do

13spacemen commented:
@Mooshimi stops the Ai's posibrain from pinging ghosts upon creation

Timberpoes reviewed (changes requested):
Your PR body is poor and does not explain what the bug is and how your proposed fix actually fixes said bug.

You have no changelog. Changelogs are expected for any change that is "player facing", but in practice we usually document all "user facing" changes too.

I politely ask you update your PR body with the necessary information and strongly advise you to read the key parts of the contrib guidelines below and utilise our template, even for minor PRs.

https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#pull-request-process

> You are going to be expected to document all your changes in the pull request. Failing to do so will mean delaying it as we will have to question why you made the change.

> We ask that you use the changelog system to document your player facing changes, which prevents our players from being caught unaware by said changes

> After leaving reviews on an open pull request, maintainers may convert it to a draft. Once you have addressed all their comments to the best of your ability, feel free to mark the pull as Ready for Review again.

We have a standard form PR template that we ask and expect contributors to follow. Most good Git clients will retrieve it for you. If your client doesn't, I suggest making the pull request direct from your repo's branch on the GitHub website.

```
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
add: Added new mechanics or gameplay changes
add: Added more things
del: Removed old things
qol: made something easier to use
balance: rebalanced something
fix: fixed a few things
soundadd: added a new sound thingy
sounddel: removed an old sound thingy
imageadd: added some icons and images
imagedel: deleted some icons and images
spellcheck: fixed a few typos
code: changed some code
refactor: refactored some code
config: changed some config setting
admin: messed with admin stuff
server: something server ops should know
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->
```

13spacemen committed:
fixes MMIs not transferring brainmob to ejected brain
adds ability to (un)anchor AIs with a wrench

Dorsisdwarf commented:
Shouldnt unanchoring ais with a wrench be tagged addition, not fix? I don't think we count "Well a different codebase has this feature so not having it is a bug"

13spacemen commented:
@Dorsisdwarf It's a bug because I wrote the AI deconstruction steps with that in mind, when you try to deconstruct an AI and its unanchored it tells you to "bolt it down first"
It's not really a balance thing because if you can get up next to an AI you can instantly card it or just kill it. Also, when an AI dies, it unanchors itself automatically anyway

Dorsisdwarf commented:
> @Dorsisdwarf It's a bug because I wrote the AI deconstruction steps with that in mind, when you try to deconstruct an AI and its unanchored it tells you to "bolt it down first". I wasn't aware being able to wrench AIs wasn't on TG It's not really a balance thing because if you can get up next to an AI you can instantly card it or just kill it. Also, when an AI dies, it unanchors itself automatically anyway

Fair enough!

Mothblocks commented:
@Timberpoes Is this okay now

Mothblocks commented on code/game/objects/structures/ai_core.dm:
```diff
@@ -72,7 +72,7 @@
 	circuit = new(src)
 	if(posibrain)
-		core_mmi = new/obj/item/mmi/posibrain(src)
+		core_mmi = new/obj/item/mmi/posibrain(src, FALSE)
```
Named parameter. Because this is new you might have to do it like:

```suggestion
		core_mmi = new/obj/item/mmi/posibrain(src, /* whatever_its_called = */ FALSE)
```

Mothblocks commented on code/modules/mob/living/silicon/ai/ai.dm:
```diff
@@ -398,7 +398,7 @@
 	var/mmi_type
 	if(posibrain_inside)
-		mmi_type = new/obj/item/mmi/posibrain(src)
+		mmi_type = new/obj/item/mmi/posibrain(src, FALSE)
```
Same

LemonInTheDark commented on code/modules/mob/living/silicon/ai/ai_defense.dm:
```diff
@@ -69,6 +69,19 @@
 	to_chat(src, span_warning("[user] shorts out your access panel lock!"))
 	emagged = TRUE
 
+/mob/living/silicon/ai/wrench_act(mob/living/user, obj/item/tool)
+	. = ..()
+	if(user.combat_mode)
+		return
+	balloon_alert(user, "[!is_anchored ? "tightening" : "loosening"] bolts...")
+	balloon_alert(src, "bolts being [!is_anchored ? "tightened" : "loosened"]...")
+	if(!tool.use_tool(src, user, 4 SECONDS))
+		return TOOL_ACT_TOOLTYPE_SUCCESS
+	flip_anchored()
+	balloon_alert(user, "bolts [is_anchored ? "tightened" : "loosened"]")
+	balloon_alert(src, "bolts [is_anchored ? "tightened" : "loosened"]")
+	return TOOL_ACT_TOOLTYPE_SUCCESS
```
Uh, why?

13spacemen replied:
@LemonInTheDark 
The AI deconstruction steps tell you to "bolt it down first!" if the AI is unanchored and you attempt deconning it. When I wrote those steps I assumed you already could wrench the AI like you can on baycode, it was only when I played on the live server and saw players hitting the AI with a wrench that I realized oh TG doesn't have that yet

Its not a balance issue, like I explained to Dorsi, if you can get up close to wrench the AI you can also just card/kill it. Also the AI can toggle itself anchored again with a verb anyway.

Timberpoes commented:
If the PR body does what it says it does, the code passes review and the changelog is accurate, it can be merged.

LemonInTheDark reviewed (changes requested):
Could you explain what was changed to fix what? you're a bit unclear and it makes this pr very hard to review

13spacemen commented:
@LemonInTheDark 
1. mmi/eject_brain() checks for brain.brainmob (which is always null since the brainmob gets moved to the mmi upon insertion) instead of checking for brainmob.
2. when you deconstruct an AI with var/posibrain_core = TRUE it spawns a new posibrain in the core structure, but whenever posibrains are created they notify ghosts so they can enter, we don't want that happening with AI posibrains
3. examining an AI tells you you can (un)wrench it, attempting to deconstruct it while its unanchored tells you to bolt it down, all of this stuff assumes you can wrench the AI but you actually can't on TG yet, I assumed you could, added a wrench_act()

13spacemen committed:
adds comments for new posibrain creation
adds arg to deactivated ai core structure Initialize() to skip creating an MMI which will be overwritten anyway

magatsuchi commented on code/game/objects/structures/ai_core.dm:
```diff
@@ -68,15 +68,16 @@
-		core_mmi = new(src)
-		core_mmi.brain = new(core_mmi)
-		core_mmi.update_appearance()
+	if(!skip_mmi_creation)
```
consider doing an early return

```
circuit = new(src)

if(skip_mmi_creation)
    return

[ your cool code here! ]

magatsuchi replied:
https://github.com/tgstation/tgstation/blob/master/.github/guides/STYLE.md#use-early-returns

thgvr commented:
Where's the "Why it's good for the game" about letting people unwrench AIs?

LemonInTheDark commented:
Isn't unbolting from the floor sort of a way for ais to "consent" to something? taking that away from them seems like something of note

13spacemen commented:
@LemonInTheDark 
If someone is right next to you and you're an AI, they can:
card you and instakill you by wiping you, or just space you
whack you to death (which will unanchor you anyway)
forcefully deconstruct you (and then they can unwrench your inactive core anyway)

So given those, I don't think it's a huge deal to allow people to unwrench AIs with wrenches. If the AI doesn't want to go they can just toggle bolts with their verb, and they can also bolt you in or stun/laser you with their turrets.

LemonInTheDark commented on code/game/objects/structures/ai_core.dm:
```diff
@@ -68,11 +68,13 @@
+		return
 	if(posibrain)
-		core_mmi = new/obj/item/mmi/posibrain(src)
+		core_mmi = new/obj/item/mmi/posibrain(src, /* autoping = */ FALSE)
```
are you embedding a comment here because new doesn't accept named arguments or something

LemonInTheDark commented on code/modules/mob/living/silicon/ai/ai.dm:
```diff
@@ -389,7 +389,7 @@
 	disconnect_shell()
 	ShutOffDoomsdayDevice()
-	var/obj/structure/ai_core/deactivated/ai_core = new(get_turf(src))
+	var/obj/structure/ai_core/deactivated/ai_core = new(get_turf(src), /* skip_mmi_creation = */ TRUE)
```
ditto

timothymtorres replied:
Did this actually work?  Think it might be causing #68567

LemonInTheDark replied:
it should do?

LemonInTheDark commented on code/modules/mob/living/silicon/ai/ai.dm:
```diff
@@ -398,7 +398,7 @@
 	var/mmi_type
 	if(posibrain_inside)
-		mmi_type = new/obj/item/mmi/posibrain(src)
+		mmi_type = new/obj/item/mmi/posibrain(src, /* autoping = */ FALSE)
```
ditto
